### ISSUE TITLE
How-to set up OAuth endpoint using Spring Boot

### ISSUE THREAD

**Issue by nickmelis** [slackapi/java-slack-sdk#555]
The documentation describes how to set up a simple Spring Boot app, however there is no clear guide on how to prepare the same Spring Boot app for distribution by setting up OAuth endpoints. I'm having to go through code and tests and can't figure out how to get it to work. It would be nice to have a comprehensive guide on how to do it, and what environment variables are required.

If this guide already exists, please let me know, as I'm currently stuck with this.

**Comment by seratch** [comment]
@nickmelis Hi, thank you for taking the time to try this SDK!

Indeed, the documents don't cover OAuth flow support in Spring Boot apps. That said, implementing it is quite simple.

* Configure your `App` to handle OAuth flow - having client_id, client_secret, scope, and so on
  * Refer to [this document](https://slack.dev/java-slack-sdk/guides/app-distribution) for details
* Have three web controllers in your Sprint Boot app - they handle:
  * POST `/slack/events` (you can customize this path)
  * GET `/slack/oauth/install` (you can customize this path)
  * GET `/slack/oauth/callback` (you can customize this path)
* Inject a singleton `App` instance into the web controllers
  * Have [`App` instance provider](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt-spring-boot-examples/src/main/java/example/SlackApp.java#L45-L46)
  * Inject the `App` instance into three web controllers via [constructor injection](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt-spring-boot-examples/src/main/java/example/SlackAppController.java#L10)

If you need a minimal working example, I can provide it early next week.

**Comment by nickmelis** [comment]
Hi @seratch, thanks for your feedback and prompt response. I am indeed a Java geek, so this SDK was my #1 choice. I think I get what you're suggesting. I'll try it next Monday, and if I can get it to work, I'm more than happy to provide code examples for the community. As a user, I feel it's all fine up to the point where you try to distribute your Slack app.

One more think I'm not clear about, when you're configuring your app in the Slack API pages, your redirect URL would then be `https://my.url/slack/oauth/callback` ? Or should I add both callback and install?

**Comment by seratch** [comment]
>when you're configuring your app in the Slack API pages, your redirect URL would then be https://my.url/slack/oauth/callback ?

Yes, the Redirect URL is `https://my.url/slack/oauth/callback`. You don't need to set `/slack/oauth/install` in any Slack app configuration pages. The `/slack/oauth/install` is the entry point that just appends a valid `state` parameter value to Slack authorize URL.

**Comment by nickmelis** [comment]
So, I've done exactly what you said. I have `App` defined as a Spring `@Bean`. Set up three `@WebServlet` controllers, and provided the following environment variables:
```
SLACK_APP_CLIENT_ID=copied from App>Basic Information>App Credentials
SLACK_APP_CLIENT_SECRET=copied from App>Basic Information>App Credentials
SLACK_APP_OAUTH_CALLBACK_PATH=/callback
SLACK_APP_OAUTH_CANCELLATION_URL=https://my.url/slack/oauth/completion
SLACK_APP_OAUTH_COMPLETION_URL=https://my.url/slack/oauth/cancellation
SLACK_APP_OAUTH_START_PATH=/start
SLACK_APP_REDIRECT_URI=https://my.url/slack/oauth/callback
SLACK_APP_SCOPE=chat:write,users:read,users:read.email
SLACK_BOT_TOKEN=copied from App>OAuth & Permissions>OAuth Tokens & Redirect URLs
SLACK_SIGNING_SECRET=copied from App>Basic Information>App Credentials
```

Now all the requests fail (including the ones that used to work before in my own workspace) with the following message in the log:

```
2020-09-05 07:57:52.019 DEBUG 18350 --- [nio-5000-exec-1] com.slack.api.bolt.App                   : Applying a middleware (name: com.slack.api.bolt.middleware.builtin.SSLCheck)
2020-09-05 07:57:52.020 DEBUG 18350 --- [nio-5000-exec-1] com.slack.api.bolt.App                   : Applying a middleware (name: com.slack.api.bolt.middleware.builtin.RequestVerification)
2020-09-05 07:57:52.020 DEBUG 18350 --- [nio-5000-exec-1] c.s.a.a.SlackSignature$Verifier          : Request verification (timestamp: 1599289071, body: payload=[...] signature: v0=0a976022e9419140934e6fb6cf41c18b5f0e7dc71621a144e2377b072fb7ea37)
2020-09-05 07:57:52.020 DEBUG 18350 --- [nio-5000-exec-1] com.slack.api.bolt.App                   : Applying a middleware (name: com.slack.api.bolt.middleware.builtin.MultiTeamsAuthorization)
2020-09-05 07:57:52.021  INFO 18350 --- [nio-5000-exec-1] c.s.a.b.m.b.MultiTeamsAuthorization      : auth.test result: null, io error: null, api error: null
```

**Comment by seratch** [comment]
MultiTeamsAuthorization tries to fetch the corresponding access token for each incoming request from Slack [this way](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/middleware/builtin/MultiTeamsAuthorization.java#L116-L120). 

When you run a Bolt app only for a single workspace, you use only a single token. This is the one you give by `SLACK_BOT_TOKEN` env variable when starting the app. By contrast, if your app supports OAuth flow, identifying a valid token for each installed workspace when receiving requests at `/slack/events` is required. To do this, you use [InstallationService](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/service/InstallationService.java) in Bolt for Java.

If you don't specify any `InstallationService`, the built-in [FileInstallationService](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/service/builtin/FileInstallationService.java) will be used. The service scans a directory to find installation information. To verify how it works, complete the OAuth flow (=Slack app installation flow with your workspace) first. By doing so, the installation data will be stored in your local file system. Then, your MultiTeamsAuthorization can lookup them.

For your production app, please implement your own InstallationService that looks up the authorization results. You can attach your InstallationService via [this setter method](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/App.java#L743) when initializing `App`.

**Comment by nickmelis** [comment]
Ok I tried to follow your instructions above. So I uninstalled the Slack app from my workspace and tried to install it again, to go through the OAuth flow. I did not specifically implement an InstallationService, and I verified the FileInstallationService is enabled by default and some (empty) folders have been created on my file system. So this is what's happening now:

- If I run with the configuration above (I had to change `SLACK_APP_OAUTH_CALLBACK_PATH` from `/callback` to `/slack/oauth/callback` to get it to work), the result is that my browser gets stuck at `https://my.url/slack/oauth/completion`, with a big 404 error. I guess this is because there's no handler for `slack/oauth/completion`? Do I have to add that one too?

- if I remove `SLACK_APP_OAUTH_CANCELLATION_URL` and `SLACK_APP_OAUTH_COMPLETION_URL` from the config above and re-run the whole process, this happens: 
```
2020-09-07 09:55:40.783 DEBUG 29115 --- [nio-5000-exec-3] com.slack.api.bolt.App                   : The handler started (request type: OAuthCallback)
2020-09-07 09:55:40.783  WARN 29115 --- [nio-5000-exec-3] .b.s.b.o.d.OAuthDefaultStateErrorHandler : Invalid state parameter detected - payload: VerificationCodePayload(code=***, state=, error=null)
2020-09-07 09:55:40.783 DEBUG 29115 --- [nio-5000-exec-3] com.slack.api.bolt.App                   : The handler completed (request type: OAuthCallback)
```
and my browser gets stuck at `https://my.url/slack/oauth/callback?code=***&state=`

What am I missing here?

**Comment by seratch** [comment]
The issue here is the lack of the `state` parameter in the redirection.

> Invalid state parameter detected - payload: VerificationCodePayload(code=***, state=, error=null)

This error message means the redirection from the Slack OAuth page does not have a valid `state` parameter. Due to that, the installation is not yet completed in this case. You can find the `/slack/oauth/callback` responds to the redirection with 40x HTTP status. As mentioned [here](https://api.slack.com/authentication/oauth-v2#asking), the state parameter should be used to avoid forgery attacks. A non-static (unique) parameter value needs to be set for each OAuth flow. 

The `SLACK_APP_OAUTH_START_PATH` endpoint (default: `/slack/oauth/start`) generates a valid state parameter for your end-users. So, please let your end-users access `https://my.url/slack/oauth/start` instead of `https://slack.com/oauth/v2/authorize?client_id=111.222...` directly. At `https://my.url/slack/oauth/start`, the server-side [redirects end-users with a valid state parameter](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/App.java#L855-L856). 

As with `/slack/oauth/callback`, the `slack/oauth/start` internally uses a service ([OAuthStateService](https://github.com/slackapi/java-slack-sdk/blob/v1.1.3/bolt/src/main/java/com/slack/api/bolt/service/OAuthStateService.java)), and you can customize the logic by implementing the interface. The default implementation uses a local file system.

>My browser gets stuck at https://my.url/slack/oauth/completion, with a big 404 error. I guess this is because there's no handler for slack/oauth/completion? Do I have to add that one too?

Yes, hosting a completion page for your end-users is recommended. The page can be a either a page served by your Spring Boot app or a simple static web page. If you don't want to have such, it's also fine to simply use Slack deep links such as `slack://app?team={team_id}&id={app_id}` or `slack://open`. https://api.slack.com/reference/deep-linking

**Comment by nickmelis** [comment]
> So, please let your end-users access https://my.url/slack/oauth/start instead of https://slack.com/oauth/v2/authorize?client_id=111.222... directly.

How do I do that?

**Comment by seratch** [comment]
Having a button linked to the URL `https://my.url/slack/oauth/start` on a web page is an option. Or, if your app is used only in multiple workspaces in an organization, just sharing the URL with installers is enough.

```
<a href="https://my.url/slack/oauth/start"><img alt="Add to Slack" height="40" width="139" src="https://platform.slack-edge.com/img/add_to_slack.png" srcSet="https://platform.slack-edge.com/img/add_to_slack.png 1x, https://platform.slack-edge.com/img/add_to_slack@2x.png 2x" /></a>
```

Also, if your app has the link in the `SLACK_APP_OAUTH_CANCELLATION_URL` page, it can be easier to navigate end-users to the right path even when a user didn't follow the right steps.

**Comment by nickmelis** [comment]
So basically, instead of using the code from the "Manage distribution" page for my Slack app (`https://app.slack.com/app-settings/TSF******/A01********/distribute`,  which is as follows:
```
<a href="https://slack.com/oauth/v2/authorize?client_id=***&scope=chat:write,users:read,users:read.email&user_scope="><img alt="Add to Slack" height="40" width="139" src="https://platform.slack-edge.com/img/add_to_slack.png" srcSet="https://platform.slack-edge.com/img/add_to_slack.png 1x, https://platform.slack-edge.com/img/add_to_slack@2x.png 2x" /></a>
```
I should modify this code to include the URL `https://my.url/slack/oauth/start` instead of `https://slack.com/oauth/v2/authorize`?

**Comment by nickmelis** [comment]
Can someone provide a minimal working example please, together with config variables to set when distributing the app? 
I'm sure I'm not the first one using this SDK, other users must have solved this problem before me.

**Comment by seratch** [comment]
@nickmelis Here is a complete example. Some information should be clearly mentioned in the documents. I will update them later. https://gist.github.com/seratch/2ae0cac0e5acf6fe6f2912000e4dbacd

**Comment by nickmelis** [comment]
@seratch do you have an answer to my earlier questions? 
- Why do I need to force people directly to `https://my.url/slack/oauth/start` rather than the standard `https://slack.com/oauth/v2/authorize`
- How to extract team ID from the completion page, as that's the only way I can make a deep link like `slack://app?team={team_id}&id={app_id}` suggested above work

**Comment by seratch** [comment]
@nickmelis 
>Why do I need to force people directly to https://my.url/slack/oauth/start rather than the standard https://slack.com/oauth/v2/authorize

The `/oauth/start` endpoint generates a `state` parameter for an end-user. The reason we strongly recommend using the state parameter is to protect your users from CSRF attacks described in OAuth 2.0 RFC: https://tools.ietf.org/html/rfc6749#section-10.12

>How to extract team ID from the completion page, as that's the only way I can make a deep link like slack://app?team={team_id}&id={app_id} suggested above work

Your callback method can get the two values in [the `oauth.v2.access` response](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/slack-api-client/src/main/java/com/slack/api/methods/response/oauth/OAuthV2AccessResponse.java).

Implement your own [OAuthV2SuccessHandler](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/bolt/src/main/java/com/slack/api/bolt/service/builtin/oauth/OAuthV2SuccessHandler.java) and set it to `App` by using [this setter method](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/bolt/src/main/java/com/slack/api/bolt/App.java#L758-L761). [This](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/bolt/src/main/java/com/slack/api/bolt/service/builtin/oauth/default_impl/OAuthV2DefaultSuccessHandler.java) is the default implementation.

**Comment by nickmelis** [comment]
@seratch just one more clarification. You said you strongly suggested using the state parameter, which is fair enough. But as far as I can see, the SDK doesn't even work without that `state` parameter. And all the links to install the app in the slack pages (website) point to `https://slack.com/oauth/v2/authorize?client_id=***&scope=chat:write,users:read,users:read.email&user_scope=`
My app is not in the marketplace yet, but I guess once it is, whoever clicks on it to install it, will end up with the same problem? 
Hope my question makes sense.

**Comment by seratch** [comment]
As long as your users click only the URLs shared by you, there is no issue. If you really need an option to bypass state parameter validation, I'm fine to add the option to this SDK. Although it's not recommended, I do understand it is useful for other users too.

**Comment by nickmelis** [comment]
Out of curiosity, am I the first one who came up with these problems? What is the other SDK's approach to this?

Just to be specific, this is the app I'm talking about: https://slack.com/apps/A019X691HCN-checkin-dev. Not sure you can access the page (you should be able to), but if you click on the green "Add to Slack" button, you'll see what I mean. 
And being this in Slack app directory, maybe we need a way to bypass state validation?

**Comment by seratch** [comment]
>Out of curiosity, am I the first one who came up with these problems? What is the other SDK's approach to this?

Node Slack SDK also requires implementing the `state` parameter. We haven't got this feedback for this SDK. I think the reasons would be 1) using the state parameter is not a blocker (I will explain below) 2) this SDK is a relatively new one (the initial version was released in March).

> if you click on the green "Add to Slack" button, you'll see what I mean. And being this in Slack app directory, maybe we need a way to bypass state validation?

Once you enable the distribution of the app, you can customize the URL of the button. Check **Settings** > **Basic Information**. If you don't find this yet, go to **Settings** > **Manage Distribution** to turn on the app distribution.

<img width="400" src="https://user-images.githubusercontent.com/19658/93211476-443c9000-f79c-11ea-895d-92f02a2782c8.png">

So, it's not a blocker to use `/oauth/start` or whatever for the installation URL. Does this make sense to you?

**Comment by nickmelis** [comment]
Ok thanks @seratch , that does answer my question. I can provide a "custom" URL to `https://my.url/slack/oauth/start` from either my landing page and the app directory. Which allows me to get around the problem I'm talking about.

**Comment by seratch** [comment]
I'm glad to hear that and noticed our documents can be improved for developers. It seems to be all clear now. Would you mind closing this issue now?

**Comment by nickmelis** [comment]
Absolutely. And I can confirm the redirection to Slack's workspace and app also works beautifully! Thanks for your help, and if you need anything from my side or have any question, please feel free to reach out!

**Comment by nickmelis** [comment]
@seratch can you see any reason why this should happen? 

```
Sep 16 08:49:53 ip-172-31-12-21 web: 2020-09-16 08:49:53.082  WARN 3652 --- [nio-5000-exec-3] .b.s.b.o.d.OAuthDefaultStateErrorHandler : Invalid state parameter detected - payload: VerificationCodePayload(code=1341368771651.1366875930595.ed0fcd5bd0596eac4cbf2fd028987fad5a54b58612103afc5eaf33330739e1ce, state=9da5ccf5-3ef6-4644-92af-20563761dfd2, error=null)
```

To give you some context, this only happens when my app is deployed on AWS. I double and triple checked all the config variables so I'm sure they're right. It doesn't happen when running locally on my machine. Also I can see the state is not null, so it contains a value. Is it perhaps just the wrong value?

**Comment by seratch** [comment]
@nickmelis Do you still use the default file-based service? If so, it works only with a single EC2 instance on AWS. If you use other runtime environments such as AWS Lambda and ECS, it doesn't work. 

Consider using [AmazonS3OAuthStateService](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/bolt/src/main/java/com/slack/api/bolt/service/builtin/AmazonS3OAuthStateService.java) or your own implementation. You can set a service instance using `App#service(OAuthStateService)` method.
https://oss.sonatype.org/service/local/repositories/releases/archive/com/slack/api/bolt/1.1.4/bolt-1.1.4-javadoc.jar/!/index.html#package

**Comment by nickmelis** [comment]
So, I'm using elastic beanstalk, which is running one single EC2 instance. So technically, it should work.
Out of curiosity, what is the default `OAuthStateService` if I don't specify one? I've just realised I am not specifying one.

**Comment by seratch** [comment]
> Out of curiosity, what is the default OAuthStateService if I don't specify one? I've just realised I am not specifying one.

It is [FileOAuthStateService](https://github.com/slackapi/java-slack-sdk/blob/v1.1.4/bolt/src/main/java/com/slack/api/bolt/service/builtin/FileOAuthStateService.java)

**Comment by nickmelis** [comment]
@seratch I have the same problem when using `AmazonS3OAuthStateService`

```
Sep 16 11:09:48 ip-172-31-12-21 web: 2020-09-16 11:09:48.996  WARN 17621 --- [nio-5000-exec-7] .b.s.b.o.d.OAuthDefaultStateErrorHandler : Invalid state parameter detected - payload: VerificationCodePayload(code=1341368771651.1390793891856.8c36c12c6d3bd17b26d13c88a187216c6bcacd4a2acfc2579e829e2bb00259a1, state=3bb132e9-4fd7-48f7-8015-e4c8819ac8bb, error=null)```

Any idea what's wrong?

**Comment by nickmelis** [comment]
Ok @seratch I think I managed to figure out what's wrong. It's to do with cookies between the `start` and `authenticate` phase, and more specifically the way the cookie `slack-app-oauth-state` is retrieved.

So, going through the `start` endpoint, a `state` is generated, saved in the app's data store and returned as a cookie with the key `slack-app-oauth-state`. This bit all works fine.

When doing the authentication bit, the `state` is retrieved from the cookies here: https://github.com/slackapi/java-slack-sdk/blob/a973bb43db9c1b6136873f3129252dba63315110/bolt/src/main/java/com/slack/api/bolt/service/OAuthStateService.java#L116
In my case, `List<String> cookieHeaders` is a list with one single element: `__cfduid=d7fabf2f46e68f2c924d91c8e20c79dad1599067038; _ga=GA1.2.659462364.1599067361; _hjid=63c4ac33-b521-4e5d-a2cd-6f9725c6e1a5; _fbp=fb.1.1599067361660.1577220329; slack-app-oauth-state=90a48a33-2356-4f32-bba9-5189c68e8a1f`

So as you can see, the code in OAuthStateService will not work and will not be able to extract the appropriate value for `state` from this string.

**Comment by nickmelis** [comment]
Let me know if you want me to raise it as a bug.

**Comment by seratch** [comment]
Thanks. We can definitely improve this. You don't need to create another issue for it.

If you're interested in fixing this, I'm happy to review your PR 👋 Otherwise, although I'm a bit busy with other tasks right now, I will be able to work on it soon.

**Comment by nickmelis** [comment]
I'm happy to submit a PR. Just need to check out the repo and get it up and running. The fix seems pretty straightforward, and I'm sure there's already tests for it. May be able to do it towards the end of the week though if that's ok.

**Comment by nickmelis** [comment]
@seratch here's the PR: https://github.com/slackapi/java-slack-sdk/pull/560
Let me know what you think

**Comment by seratch** [comment]
@nickmelis Thanks for submitting the PR. I really appreciate it. Please check the suggestions (just formatting issues) and the CLA.

**Comment by seratch** [comment]
@nickmelis Thanks for taking the time to make a fix for the bug. I just released a new patch version - 1.1.5. Try it out! https://github.com/slackapi/java-slack-sdk/releases/tag/v1.1.5
